phoenix-core/src/main/java/org/apache/phoenix/execute/MutationState.java

```diff
@@ -931,220 +944,323 @@ private void send(Iterator<TableRef> tableRefIterator) throws SQLException {
-                                    sqlE.setNextException(ServerUtil.parseServerException(e));
-                                } else {
-                                    sqlE = ServerUtil.parseServerException(e);
-                                }
+                            hTable.close();
+                        } catch (IOException e) {
+                            if (sqlE != null) {
+                                sqlE.setNextException(ServerUtil.parseServerException(e));
+                            } else {
+                                sqlE = ServerUtil.parseServerException(e);
                             }
-                            if (sqlE != null) { throw sqlE; }
                         }
+                        if (sqlE != null) { throw sqlE; }
                     }
-                } while (shouldRetry && retryCount++ < 1);
+                }
+            } while (shouldRetry && retryCount++ < 1);
+        }
+    }
+
+    private Map<TableInfo, List<Mutation>> filterIndexCheckerMutations(Map<TableInfo, List<Mutation>> mutationMap,
+            Map<TableInfo, List<Mutation>> deletedIndexMutations) {
+        Map<TableInfo, List<Mutation>> verifiedMutationsMap = new LinkedHashMap<>();
+        Iterator<Entry<TableInfo, List<Mutation>>> mapIter = mutationMap.entrySet().iterator();
+        while (mapIter.hasNext()) {
+            Entry<TableInfo, List<Mutation>> pair = mapIter.next();
+            TableInfo tableInfo = pair.getKey();
+            if (IndexUtil.isGlobalIndexCheckerEnabled(connection, tableInfo.hTableName)) {
+                PTable table = tableInfo.getOrigTableRef().getTable();
+                byte[] emptyCF = SchemaUtil.getEmptyColumnFamily(table);
+                byte[] emptyCQ = EncodedColumnsUtil.getEmptyKeyValueInfo(table).getFirst();
+
+                List<Mutation> mutations = pair.getValue();
+
+                if (mutations.size() == 0) {
+                    continue;
+                }
+
+                List<Mutation> indexMutations = new LinkedList<>();
+                for (Mutation m : mutations) {
+                    if (m instanceof Delete) {
+                        Put put = new Put(m.getRow());
+                        put.addColumn(emptyCF, emptyCQ, m.getTimestamp(), FALSE_BYTES);
+                        indexMutations.add(put);
+                        if (deletedIndexMutations.containsKey(tableInfo)) {
+                            deletedIndexMutations.get(tableInfo).add(m);
+                        } else {
+                            List<Mutation> mts = Lists.newArrayList();
+                            mts.add(m);
+                            deletedIndexMutations.put(tableInfo, mts);
+                        }
+                    }
+                }
+
+                verifiedMutationsMap.put(pair.getKey(), indexMutations);
+                mapIter.remove();
+            }
+
+        }
+        return verifiedMutationsMap;
```

Comment:
If I understood this function correctly, I find the variable name little misleading. verifiedMutationMap --> mutations in the map are verified, which doesn't seem to be the case. Rather, the mutations are from the first phase where verified column is set to false.  Please correct me if I am wrong.

Reply:
I see why it is confusing, I will do some more refactoring here